ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,536 @@
+    ZeroKnowledgeProof<F, R>: ProofArray,
+{
+    fn len(&self) -> usize {
+        self.proofs.len() * <ZeroKnowledgeProof<F, R> as ProofArray>::Length::USIZE
+    }
+
+    fn new() -> Self {
+        ProofBatch { proofs: Vec::new() }
+    }
+
+    /// This function sends a `ProofBatch` to the party on the left
+    /// and receives a `ProofBatch` from the party on the right.
+    ///
+    /// It sets `total_records` in the input context to the `ProofBatch` length,
+    /// i.e. it expects to receive a `ProofBatch` of the same length as being sent.
+    async fn send_and_receive<C>(&self, ctx: C) -> Self
+    where
+        C: Context,
+    {
+        // set up context
+        let ctx_left = ctx.set_total_records(self.len());
+
+        // set up channels
+        let send_channel_left = &ctx_left.send_channel::<F>(ctx.role().peer(Direction::Left));
+        let receive_channel_right = &ctx_left.recv_channel::<F>(ctx.role().peer(Direction::Right));
+
+        // prepare variable for received messages
+        let mut receive_right = ProofBatch::new();
+
+        // send to left, receive on the right
+        // we send the proof batch via sending the individual field elements
+        seq_join(
+            ctx_left.active_work(),
+            stream::iter(self.proofs.iter().flat_map(|x| x.g.iter()).enumerate().map(
+                |(i, &x)| async move {
+                    send_channel_left.send(RecordId::from(i), x).await.unwrap();
+                    receive_channel_right
+                        .receive(RecordId::from(i))
+                        .await
+                        .unwrap()
+                },
+            )),
+        )
+        //.collect::<Vec<GenericArray<F, N>>>()
+        .collect::<Vec<F>>()
+        .await
+        .chunks(<ZeroKnowledgeProof<F, R> as ProofArray>::Length::USIZE)
+        .for_each(|x| {
+            receive_right.proofs.push(ZeroKnowledgeProof {
+                g: GenericArray::<F, <ZeroKnowledgeProof<F, R> as ProofArray>::Length>::from_slice(
+                    x,
+                )
+                .clone(),
+            });
+        });
+
+        // output
+        receive_right
+    }
+
+    /// This function generates a batch of quadratic proofs
+    /// it iteratively generates `ZeroKnowledgeProofs` and collects them in a batch
+    ///
+    /// It generates `prover_left_batch` as its left output
+    /// and `verifier_left_batch` as its right output.
+    /// Further, it outputs the generated masks included in the last proof of the batch.
+    ///
+    /// `PRSS` is invoked since `PRSS` right is used to generate `prover_left_batch`.
+    /// Since `PRSS` cannot be invoked later again on the same `Gate` and `Index`,
+    /// we use `PRSS` left to generate `verifier_left_batch` and output it as well.
+    ///
+    /// Outputs `(prover_left_batch, verifier_left_batch, (p_mask, q_mask))`
+    fn compute_proof_batch<C, I>(ctx: &C, uv_tuple_block: I) -> (Self, Self, (F, F))
```

Comment:
No, because the prover_left_batch needs to be exchanged to get verifier_right_batch. Using VerifierBatch that consists of verifier_left_batch, verifier_right_batch would be confusing here.